File: concierge_packages/installer/src/install_concierge/installer_lib/do_install.py
import subprocess
import os
import shutil
from script_builder.argument_processor import ArgumentProcessor
from isi_util.password import get_strong_password, generate_strong_password
from importlib.metadata import version
from .package_dir import package_dir
from .docker_compose_helper import docker_compose_helper
import requests
from concierge_util import load_config, write_config
from .set_env import set_env
from .docker_containers import keycloak_exists
from .create_certificates import create_certificates


def do_install(
    argument_processor: ArgumentProcessor, environment="production", is_local=False
):
    # the development environment uses different docker compose files which should already be in the cwd
    if environment != "development":
        # for production we need to copy the compose files from the package into the cwd because docker compose reads the .env file from the same directory as the launched files
        shutil.copytree(
            os.path.join(package_dir, "docker_compose"), os.getcwd(), dirs_exist_ok=True
        )
    config = load_config()
    # Keycloak needs the host and port to be set before as these are piped into the initial realm config
    set_env("WEB_HOST", argument_processor.parameters["host"])
    set_env("WEB_PORT", str(argument_processor.parameters["port"]))
    if argument_processor.parameters["security_level"] != "None":
        set_env("CONCIERGE_SECURITY_ENABLED", "True")
        set_env("CONCIERGE_SERVICE", "concierge-enable-security")
        # TODO: handle existing certificates
        # TODO: handle certificates supplied by the user
        cert_dir = os.path.join(os.getcwd(), "self_signed_certificates")
        create_certificates(cert_dir)
        set_env("ROOT_CA", os.path.join(cert_dir, "root-ca.pem"))
        set_env(
            "OPENSEARCH_CLIENT_CERT",
            os.path.join(cert_dir, "opensearch-admin-client-cert.pem"),
        )
        set_env(
            "OPENSEARCH_CLIENT_KEY",
            os.path.join(cert_dir, "opensearch-admin-client-key.pem"),
        )
        set_env(
            "OPENSEARCH_ADMIN_KEY", os.path.join(cert_dir, "opensearch-admin-key.pem")
        )
        set_env(
            "OPENSEARCH_ADMIN_CERT", os.path.join(cert_dir, "opensearch-admin-cert.pem")
        )
        set_env(
            "OPENSEARCH_NODE_CERT", os.path.join(cert_dir, "opensearch-node1-cert.pem")
        )
        set_env(
            "OPENSEARCH_NODE_KEY", os.path.join(cert_dir, "opensearch-node1-key.pem")
        )
        set_env("KEYCLOAK_CERT", os.path.join(cert_dir, "keycloak-cert.pem"))
        set_env("KEYCLOAK_CERT_KEY", os.path.join(cert_dir, "keycloak-key.pem"))
        set_env("WEB_CERT", os.path.join(cert_dir, "concierge-cert.pem"))
        set_env("WEB_KEY", os.path.join(cert_dir, "concierge-key.pem"))
        # we only configure Keycloak if it wasn't already present
        if not keycloak_exists():
            keycloak_password = get_strong_password(
                "Enter password for initial Keycloak admin account: "
            )
            db_password = generate_strong_password()
            set_env("KEYCLOAK_INITIAL_ADMIN_PASSWORD", keycloak_password)
            set_env("POSTGRES_DB_PASSWORD", db_password)
            full_path = os.path.join(os.getcwd(), "docker-compose-launch-keycloak.yml")
            # pull latest versions before launching Keycloak, otherwise the instance gets broken if new images are pulled during the next step
            subprocess.run(["docker", "compose", "-f", full_path, "pull"])
            subprocess.run(
                [
                    "docker",
                    "compose",
                    "-f",
                    full_path,
                    "up",
                    "-d",
                ]
            )
            # TODO: some kind of timeout on this?
            print(
                "Waiting for Keycloak to start so we can get the OpenID credentials..."
            )
            print("This can take a few minutes, please be patient!")
            while True:
                try:
                    response = requests.post(
                        "https://localhost:8443/realms/master/protocol/openid-connect/token",
                        {
                            "client_id": "admin-cli",
                            "grant_type": "password",
                            "username": "admin",
                            "password": keycloak_password,
                        },
                        verify=os.getenv("ROOT_CA"),
                    )
                    token = response.json()["access_token"]
                    headers = {
                        "content-type": "application/json",
                        "Authorization": f"Bearer {token}",
                    }
                    # TODO: we should get the client ID from the realm JSON file to avoid errors
                    response = requests.get(
                        "https://localhost:8443/admin/realms/concierge/clients/7a3ec428-36f2-49c4-91b1-8288dc44acb0/client-secret",
                        headers=headers,
                        verify=os.getenv("ROOT_CA"),
                    )
                    keycloak_secret = response.json()["value"]
                    break
                except Exception:
                    pass

            set_env("KEYCLOAK_CLIENT_ID", "concierge-auth")
            set_env("KEYCLOAK_CLIENT_SECRET", keycloak_secret)
        set_env("OPENSEARCH_SERVICE", "opensearch-node-enable-security")

        set_env("KEYCLOAK_SERVICE_FILE", "docker-compose-keycloak.yml")
    else:
        set_env("CONCIERGE_SERVICE", "concierge")
        set_env("CONCIERGE_SECURITY_ENABLED", "False")
        set_env("OPENSEARCH_SERVICE", "opensearch-node-disable-security")
        set_env("KEYCLOAK_SERVICE_FILE", "docker-compose-blank.yml")
    set_env("ENVIRONMENT", environment)
    set_env("CONCIERGE_VERSION", version("install_concierge"))
    set_env(
        "OLLAMA_SERVICE",
        "ollama-gpu"
        if argument_processor.parameters["compute_method"].lower() == "gpu"
        else "ollama",
    )
    write_config(config)
    # docker compose
    docker_compose_helper(environment, is_local, True)
    # ollama model load
    subprocess.run(
        [
            "docker",
            "exec",
            "-it",
            "ollama",
            "ollama",
            "pull",
            argument_processor.parameters["language_model"],
        ]
    )

File: bun_installer/server/doInstall.ts
import { $ } from "bun"
import createCertificates from "./createCertificates"
import path from "node:path"
import { keycloakExists } from "./dockerItemsExist"
import crypto from "node:crypto"
import KcAdminClient from '@keycloak/keycloak-admin-client'
import getEnvPath from "./getEnvPath"
import * as envfile from "envfile"
import getVersion from "./getVersion"
import runPython from "./runPython"
import util from "node:util"
import logMessage from "./logMessage"
import configurePreCommit from "./configurePreCommit"
const exec = util.promisify(await import("node:child_process").then(child_process => child_process.exec))

export default async function* (options: FormData) {
    const environment = options.get("dev_mode")?.toString() == "True" ? "development" : "production"
    // we keep track of the environment variables so we can write to the .env file and the process environment as needed
    const envs: {[key: string]: string} = {}
    const updateEnv = () => {
        Object.entries(envs).forEach(([key, value]) => process.env[key] = value)
        return Bun.write(getEnvPath(), envfile.stringify(envs))
    }
    envs.WEB_HOST = options.get("host")?.toString() || "localhost"
    envs.WEB_PORT = options.get("port")?.toString() || "15130"
    const securityLevel = options.get("security_level")?.toString()
    if (securityLevel && securityLevel != "none") {
        yield logMessage("configuring security...")
        envs.CONCIERGE_SECURITY_ENABLED = "True"
        envs.CONCIERGE_SERVICE = "concierge-enable-security"
        const certDir = path.resolve("self_signed_certificates")        
        await createCertificates(certDir)
        envs.ROOT_CA = path.join(certDir, "root-ca.pem")
        envs.OPENSEARCH_CLIENT_CERT = path.join(certDir, "opensearch-admin-client-cert.pem")
        envs.OPENSEARCH_CLIENT_KEY = path.join(certDir, "opensearch-admin-client-key.pem")
        envs.OPENSEARCH_ADMIN_KEY = path.join(certDir, "opensearch-admin-key.pem")
        envs.OPENSEARCH_ADMIN_CERT = path.join(certDir, "opensearch-admin-cert.pem")
        envs.OPENSEARCH_NODE_CERT = path.join(certDir, "opensearch-node1-cert.pem")
        envs.OPENSEARCH_NODE_KEY = path.join(certDir, "opensearch-node1-key.pem")
        envs.KEYCLOAK_CERT = path.join(certDir, "keycloak-cert.pem")
        envs.KEYCLOAK_CERT_KEY = path.join(certDir, "keycloak-key.pem")
        envs.WEB_CERT = path.join(certDir, "concierge-cert.pem")
        envs.WEB_KEY = path.join(certDir, "concierge-key.pem")
        yield logMessage("configured TLS certificates.")
        if (!await keycloakExists()) {
            const keycloakPassword = options.get("keycloak_password")?.toString()
            // TODO: create error type for this
            if (!keycloakPassword) throw new Error("Keycloak admin password is invalid!")
            // TODO: validate password strength
            // in case we have old postgres data we should remove it to avoid mismatches
            await $`docker volume rm --force concierge_postgres_data`.quiet()
            envs.KEYCLOAK_INITIAL_ADMIN_PASSWORD = keycloakPassword
            const postgresPassword = crypto.randomBytes(25).toString("hex")
            // TODO: validate password strength
            envs.POSTGRES_DB_PASSWORD = postgresPassword
            updateEnv()
            yield logMessage("getting OpenID credentials from Keycloak service. This can take a few minutes!")
            const keycloakComposeFile = path.join("docker_compose", "docker-compose-launch-keycloak.yml")
            // update the keycloak image otherwise it can get overwritten when we launch everything below
            await $`docker compose -f ${keycloakComposeFile} pull`
            await $`docker compose -f ${keycloakComposeFile} up -d`
            // TODO: we need to use the root CA certs rather than disabling verification!
            process.env.NODE_TLS_REJECT_UNAUTHORIZED='0'
            // keep trying this until keycloak is up
            while(true) {
                try {
                    const kcClient = new KcAdminClient({
                        baseUrl: 'https://localhost:8443'
                    })
                    await kcClient.auth({
                        username: 'admin',
                        password: keycloakPassword,
                        grantType: 'password',
                        clientId: 'admin-cli',
                      });
                    const secret = await kcClient.clients.getClientSecret({id: '7a3ec428-36f2-49c4-91b1-8288dc44acb0', realm: 'concierge'})
                    envs.KEYCLOAK_CLIENT_ID = "concierge-auth"
                    envs.KEYCLOAK_CLIENT_SECRET = secret.value!
                    break
                }  
                catch (error) {
                    continue
                }
            }
            yield logMessage("got Keycloak credentials.")
        }
        envs.OPENSEARCH_SERVICE = "opensearch-node-enable-security"
        envs.KEYCLOAK_SERVICE_FILE = "docker-compose-keycloak.yml"
    }
    else {
        envs.CONCIERGE_SERVICE = "concierge"
        envs.CONCIERGE_SECURITY_ENABLED = "False"
        envs.OPENSEARCH_SERVICE = "opensearch-node-disable-security"
        envs.KEYCLOAK_SERVICE_FILE = "docker-compose-blank.yml"
    }
    envs.ENVIRONMENT = environment
    envs.CONCIERGE_VERSION = getVersion()
    envs.OLLAMA_SERVICE = options.has("use_gpu") ? "ollama-gpu" : "ollama"
    await updateEnv()
    yield logMessage("launching Docker containers. This can take quite a long time if this is your first launch or updates have been released to the Docker images...")
    if (environment == "development") {
        await $`docker compose -f ./docker_compose/docker-compose-dev.yml pull`
        await $`docker compose -f ./docker_compose/docker-compose-dev.yml up -d`
        yield logMessage("configuring Python environment...")
        await exec("python3 -m venv ..")
        await runPython("pip install -r dev_requirements.txt")
        await configurePreCommit()
    } 
    else {
        await $`docker compose -f ./docker_compose/docker-compose.yml pull`
        await $`docker compose -f ./docker_compose/docker-compose.yml up -d`
    } 
    if (securityLevel == "demo") {
        yield logMessage("adding demo users")
        envs.IS_SECURITY_DEMO = "True"
        await updateEnv()
        if (environment == "development") await runPython("concierge_scripts.add_keycloak_demo_users")
        else await $`docker exec -d concierge python -m concierge_scripts.add_keycloak_demo_users`
    }
    yield logMessage("pulling language model. This can take quite a long time if you haven't downloaded the model before.")
    // TODO use options.get("language_model")
    await $`docker exec -it ollama ollama pull mistral`
    console.log("Installation done\n")
}